Add in-process mpsc-channels based implementation

vvuk opened the pull request:
This PR adds an in-process-only implementation on top of mpsc channels.  It uses this as the default on windows until we finish a real windows implementation.

pcwalton commented on platform/inprocess/mod.rs:
```diff
@@ -0,0 +1,359 @@
+                if h.id() == id {
+                    r_index = index;
+                    r_id = self.receiver_ids[index] as i64;
+                    //println!("r_id {}", r_index);
```
nit: remove these commented out printlns

pcwalton commented on platform/inprocess/mod.rs:
```diff
@@ -0,0 +1,359 @@
+    {
+        match self.sender.borrow().send(MpscChannelMessage(data.to_vec(), ports, shared_memory_regions)) {
+            Err(_) => { Err(MpscError::ChannelClosedError) },
+            Ok(_) => { Ok(()) },
```
nit: no need for the braces here

pcwalton commented on platform/inprocess/mod.rs:
```diff
@@ -0,0 +1,359 @@
+    }
+
+    pub fn add(&mut self, receiver: MpscReceiver) -> Result<i64,MpscError> {
+        self.last_index = self.last_index + 1;
```
nit: `self.last_index += 1;` is shorter

pcwalton commented on platform/inprocess/mod.rs:
```diff
@@ -0,0 +1,359 @@
+
+        let receivers = &mut self.receivers;
+        match receivers[r_index].recv() {
+            Ok((data, channels, shmems)) => return Ok(vec![MpscSelectionResult::DataReceived(r_id, data, channels, shmems)]),
```
nit: I don't think you need the `return` here or on any of the other branches

pcwalton commented on platform/inprocess/mod.rs:
```diff
@@ -0,0 +1,359 @@
+            Ok((data, channels, shmems)) => return Ok(vec![MpscSelectionResult::DataReceived(r_id, data, channels, shmems)]),
+            Err(err) => {
+                match err {
+                    MpscError::ChannelClosedError => {
```
Could just be folded into the match above: `Err(MpscError::ChannelClosedError) => { ... } Err(_) => { ... }`

pcwalton commented on platform/inprocess/mod.rs:
```diff
@@ -0,0 +1,359 @@
+        }
+    }
+
+    fn signal_acquisition(&self) -> () {
```
nit: no need for `-> ()`

pcwalton commented on platform/inprocess/mod.rs:
```diff
@@ -0,0 +1,359 @@
+
+    fn signal_acquisition(&self) -> () {
+        let mut notified = self.lock.lock().unwrap();
+        *notified = true;
```
Is there a reason why this lock/condvar code can't be just an MPSC `Mutex<Receiver<()>>`/`Sender<()>` pair? That is, clients just lock the receiver and wait. This code isn't performance critical and I find channels easier to think about than condvars (maybe it's just from writing a lot of Servo code, though).

vvuk replied:
In theory I'd need the sender to also block until the receiver actually finishes receiving.  But maybe that's not actually necessary in this case, since the actual connection setup can't really fail...

pcwalton commented on platform/inprocess/mod.rs:
```diff
@@ -0,0 +1,359 @@
+        let r = self.receiver.borrow();
+        match r.as_ref().unwrap().recv() {
+            Ok(MpscChannelMessage(d,c,s)) => Ok((d,
+                                                 c.into_iter().map(|x| OpaqueMpscChannel::new(x)).collect(),
```
Does `c.into_iter().map(OpaqueMpscChannel::new).collect()` work?

pcwalton commented on platform/inprocess/mod.rs:
```diff
@@ -0,0 +1,359 @@
+            for r in &self.receivers {
+                let inner_r = mem::replace(&mut *r.receiver.borrow_mut(), None);
+                receivers.push(inner_r);
+            }
```
Instead of swapping out each receiver individually, is it possible to do something like `let receivers = mem::replace(&mut self.receivers, Vec::new()).into_iter().map(Some).collect()`?

vvuk replied:
I don't think so, at least not something that simple... the loop I have is reaching replacing the MpscReceiver's RefCell<Option<>> member value's with None, and pulling out the inner mpsc::Receiver.  It's not just replacing the self.receivers array values themselves (they don't change).

I could probably write it as a map that does a mem::replace, but I don't know that it would be any clearer.

vvuk commented:
Okay, converted to using a mpsc::channel() instead of a mutex/condvar (and renamed ONE_SHOT_SENDERS to ONE_SHOT_SERVERS, derp).  Though now I'm actually confused as to how this is working at all -- OneShotServer::accept() does a get on the ONE_SHOT_SERVERS hashmap, and then calls accept() which calls recv(), which should block.

MpscSender::connect() does a remove on the ONE_SHOT_SERVERS hashmap and calls send() to poke the server.  But the server part should be blocked in recv() inside accept()... so who owns the ServerRecord?

pcwalton commented:
I'm confused—are you asking who owns the memory for the `ServerRecord`? I don't think it matters, since all of its elements are heap allocated and atomically reference counted.

vvuk commented:
Yeah, I guess I'm getting a copy during the get().  This should be good to go.
